File: 백준/Silver/15657. N과 M （8）/N과 M （8）.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Main {
	public static void main(String[] args) throws IOException {
		input();
		solution();
		output();
	}
	private static int N, M;
	private static int[] arr;
	private static List<String> list;
	private static void solution() {
		list = new ArrayList<>();
		int[] res = new int[M];
		comb(res, 0, 0);
	}
	private static void comb(int[] res, int idx, int depth) {
		if (depth == M) {
			list.add(IntStream.of(res).mapToObj(String::valueOf).collect(Collectors.joining(" ")));
			return;
		}

		for (int i = idx; i < N; i++) {
			res[depth] = arr[i];
			comb(res, i, depth + 1);
		}
	}
	private static void output() throws IOException {
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		bw.write(String.join("\n", list));
		bw.flush();
	}
	private static void input() throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int[] nm = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		arr = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).sorted().toArray();
		br.close();
		N = nm[0];
		M = nm[1];
	}
}

File: 백준/Silver/15657. N과 M （8）/N과 M （8）.kt
fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readln().split(" ").map { it.toInt() }
    val arr = readln().split(" ").map { it.toInt() }.sorted().toIntArray()
    close()

    val tmp = IntArray(m)
    comb(arr, tmp, 0, m, 0)
    println(sb.toString())
}

private fun comb(arr : IntArray, tmp : IntArray, depth : Int, m : Int, idx : Int) {
    if (depth == m) {
        toString(tmp)
        return
    }

    for (i in idx .. arr.lastIndex) {
        tmp[depth] = arr[i]
        comb(arr, tmp, depth + 1, m, i)
    }
}

private val sb = StringBuilder()
private fun toString(arr : IntArray) =
    sb.append(
        arr.joinToString(" ")
    ).append("\n")